### PR TITLE
WIP: pass context into all run methods

### PR DESCRIPTION
**Pull request by justinsb**
- Context threading: more wiring
- More context threading
- Pass context into Run methods

### CODE REVIEW

**Comment by justinsb** [comment]
I'd like to sequence this after the generics merge, but I think we will automatically pass the context in, so I want to get this under testing...

**Comment by johngmyers** [comment]
Perhaps the context should go into the Target? I'm not seeing much of a problem keeping it in the fi.Context, since I can't think of a use case for the scopes being different.

File: pkg/client/simple/vfsclientset/clientset.go [hunk 1]
```diff
@@ -64,9 +64,9 @@ func (c *VFSClientset) ListClusters(ctx context.Context, options metav1.ListOpti
 }
 
 // ConfigBaseFor implements the ConfigBaseFor method of simple.Clientset for a VFS-backed state store
-func (c *VFSClientset) ConfigBaseFor(cluster *kops.Cluster) (vfs.Path, error) {
+func (c *VFSClientset) ConfigBaseFor(ctx context.Context, cluster *kops.Cluster) (vfs.Path, error) {
 	if cluster.Spec.ConfigBase != "" {
-		return vfs.Context.BuildVfsPath(cluster.Spec.ConfigBase)
+		return vfs.FromContext(ctx).BuildVfsPath(cluster.Spec.ConfigBase)
```
Comment (johngmyers) [inline pkg/client/simple/vfsclientset/clientset.go:69]: I'm struggling to see why the `vfs.Path` should depend on a passed-in `context.Context`. And only when the `ConfigBase` is set.

Shouldn't this either take a VFSContext or hang the VFSContext off of the VFSClientset?

Comment (johngmyers) [reply pkg/client/simple/vfsclientset/clientset.go:69]: It looks like a VFSContext is closer to a Clientset than it is to a context.Context.

Comment (justinsb) [reply pkg/client/simple/vfsclientset/clientset.go:69]: > It looks like a VFSContext is closer to a Clientset than it is to a context.Context.

Yes, I think that's fair.  For example, if we had two different identities (which we don't, but if we did)... then I would expect we would have implement that with different VFSContext instances.

We could rename VFSContext - perhaps VFSScope ?

Our current use-case is for tests, so I didn't want to do _too_ much refactoring, but it prompts introducing contexts so perhaps that ship has sailed.  I do think that introducing contexts is good, even if we're not using the tracing / logging etc that it enables yet.

Comment (johngmyers) [reply pkg/client/simple/vfsclientset/clientset.go:69]: It's sort of a `VFSClient`. It's a pity we have `VFSClientSet`; perhaps that should be `VFSSimpleClientset`?

Comment (johngmyers) [reply pkg/client/simple/vfsclientset/clientset.go:69]: I don't have an objection to introducing `context.Context`s; it's just that this doesn't seem an appropriate use of them.

Comment (justinsb) [reply pkg/client/simple/vfsclientset/clientset.go:69]: That's reasonable - there's certainly an element of expedience in my thinking here.  If we add context.Context to a bunch more functions, then I think we can do it the more logical way you're suggesting.  I was trying to show some value before we added context.Context everywhere, but it sounds like we're already bought in to / resigned to the need for context.Context!

**Pull request closed without merging**